lib/iterator.js

```diff
@@ -0,0 +1,312 @@
+'use strict';
+
+const toIterator = base => {
+  if (base[Symbol.iterator]) {
+    return base[Symbol.iterator]();
+  } else if (base.next) {
+    return base;
+  } else {
+    throw new TypeError('Base is neither Iterator nor Iterable');
+  }
+};
+
+class Iterator {
+  constructor(base) {
+    this.base = toIterator(base);
+  }
+
+  [Symbol.iterator]() {
+    return this;
+  }
+
+  next() {
+    return this.base.next();
+  }
+
+  each(fn, thisArg) {
+    this.forEach(fn, thisArg);
+  }
+
+  forEach(fn, thisArg) {
+    for (const value of this) {
+      fn.call(thisArg, value);
+    }
+  }
+
+  every(predicate, thisArg) {
+    for (const value of this) {
+      if (!predicate.call(thisArg, value)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  find(predicate, thisArg) {
+    for (const value of this) {
+      if (predicate.call(thisArg, value)) {
+        return value;
+      }
+    }
+  }
+
+  includes(element) {
+    for (const value of this) {
+      if (value === element) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  reduce(reducer, initialValue) {
+    let result = initialValue;
+
+    if (result === undefined) {
+      const next = this.next();
+      if (next.done) {
+        throw new TypeError(
+          'reduce of consumed iterator with no initial value'
+        );
+      }
+      result = next.value;
+    }
+
+    for (const value of this) {
+      result = reducer(result, value);
+    }
+    return result;
+  }
+
+  some(predicate, thisArg) {
+    for (const value of this) {
+      if (predicate.call(thisArg, value)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  collectTo(CollectionClass) {
+    return new CollectionClass(this);
+  }
+
+  collectWith(obj, collector) {
+    this.forEach(element => collector(element, obj));
+  }
+
+  toArray() {
+    return Array.from(this);
+  }
+
+  map(mapper, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new MapIterator(this, mapper, thisArg);
+  }
+
+  filter(predicate, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FilterIterator(this, predicate, thisArg);
+  }
+
+  flat(depth = 1) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatIterator(this, depth);
+  }
+
+  flatMap(mapper, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatMapIterator(this, mapper, thisArg);
+  }
+
+  zip(...iterators) {
+    // eslint-disable-next-line no-use-before-define
+    return new ZipIterator(this, iterators);
+  }
+
+  join(...iterators) {
+    // eslint-disable-next-line no-use-before-define
+    return new JoinIterator(this, new Iterator(iterators));
```

Comment:
Why do you need to create `new Iterator` from `iterators` array here?

Reply:
So that I don't need to shift an array or store an index.

Reply:
iterators[Symbol.iterator] or toIterator(iterators)?

Reply:
@nechaido, but you are already converting this array to an iterator [here](https://github.com/metarhia/common/pull/123/files/28acd54c78c568489f2b40df9fad8fd0b698edd3#diff-d217736e9b27c173b1b076c6fbd6416aR290).

Reply:
And what if iterators is an Iterator itself?

Reply:
@nechaido, `iterators` is an array here.